libs/algo/odc/algo/_percentile.py

```diff
@@ -0,0 +1,58 @@
+from typing import List, Sequence, Tuple
+import dask.array as da
+import xarray as xr
+import numpy as np
+from ._geomedian import reshape_for_geomedian
+from ._masking import keep_good_np
+
+
+def np_percentile(xx, percentile, nodata):
+
+    mask = xx != nodata
+    valid_counts = mask.sum(axis=-1)
+
+    sorted_arr = np.sort(xx, axis=-1)
+    index_offsets = np.arange(0, xx.size, xx.shape[-1])
+
+    indices = np.round(percentile * (valid_counts - 1)).astype(np.int64).flatten()
+    indices = indices + index_offsets
+
+    xx = sorted_arr.take(indices).reshape(xx.shape[:-1])
+
+    return keep_good_np(xx, (valid_counts >= 3), nodata)
+
+
+def xr_percentile(
+    src: xr.Dataset,
+    percentiles: Sequence,
+    nodata,
+) -> xr.Dataset:
+
+    """
+    Calculates the percentiles of the input data along the time dimension.
+
+    :param src: xr.Dataset, bands can be either
+        float or integer with `nodata` values to indicate gaps in data.
+        `nodata` must be the largest values in the dataset or NaN.
```

Comment:
that's not feasible, and also not necessary, you can just as well deal with smallest value, and even middle value is not that much harder to remove post sorting.

Reply:
but I guess supporting floating point only is fine.

Reply:
doing computation in masked int is too hard, let's just compute in float with conversion back, it's almost always `(u)int16` so no loss of precision when computation is done in `float32`, if you want to detect `(u)int32` and use `float64` for those then that's fine too.

options for conversion are
- inside `np_percentile` a bit more annoying, but smaller Dask graph
- as a Dask operation with already existing methods (more Dask nodes, but simpler `np_percentile`)

Reply:
aw man, I was hoping to avoiding duplicating everything to caste as `floats`. Are there use cases where the `nodata` isn't the largest or smallest value in the range?

Reply:
it's usually smallest, -9999 for otherwise positive data, other common case is `0` for `uint16`, so typically it's the smallest value, but I guess `0xFFFF` can happen too on unsigned data.

Reply:
placing nodata in the middle of your valid range seems odd, haven't seen that in the wild.